### ISSUE TITLE
Cannot mount a FS in dokan...

### ISSUE THREAD

**Issue by apocalypse2012** [PyFilesystem/pyfilesystem#256]
Python 2.7.9 64 bit and 32 bit
Dokan 0.7.4
FS 0.5.4
Windows 8.1

>>> from fs.memoryfs import MemoryFS
>>> from fs.expose import dokan
>>> fs = MemoryFS()
>>> mp = dokan.mount(fs,"Q")
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Python27\lib\site-packages\fs\expose\dokan\__init__.py", line 936, in mount
    check_ready(mp)
  File "C:\Python27\lib\site-packages\fs\expose\dokan\__init__.py", line 913, in check_ready
    raise OSError("dokan mount process seems to be hung")
OSError: dokan mount process seems to be hung

>>> dokan.mount(fs,"Q",foreground=True)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Python27\lib\site-packages\fs\expose\dokan\__init__.py", line 929, in mount
    raise OSError("Dokan failed with error: %d" % (res,))
OSError: Dokan failed with error: -5

I tried the newer versions before I realized that 8.0 was API breaking. I've read through all the message boards I can find with no apparent 'fix'. Dokan 0.6.0 no longer appears to be available on the Dokan project page at Github. I

**Comment by lurch** [comment]
Somebody started working on updated Dokan support (#241 and #236) but I dunno if they ever got any further :-/
Ping @zommuter

See also https://github.com/PyFilesystem/pyfilesystem/search?q=dokan&state=open&type=Issues

**Comment by zommuter** [comment]
Thanks for the ping, @lurch. Unfortunately that is currently at the very bottom of my TODO list, and lacking more experience with Dokan I doubt I can fix this on my own :(

**Comment by Liryna** [comment]
:+1: I am willing to help on this !
@arekbulski I seen you forked pyfilesystem, did you plan to update dokan implentation ?

I will try to take a look at it in the next days and see what changes should be made.

**Comment by lurch** [comment]
It would be great to see working Dokan support in pyfilesystem again :-)

**Comment by arekbulski** [comment]
I am willing to review pyfilesystem, especially documentation. But I am also looking forward to seeing Dokan working with pyfilesystem, tell me how can I help. @Liryna

**Comment by Liryna** [comment]
@arekbulski Ok, I just wanted to know if you planned to update the wrapper yourself.

I began to update most of the natives commandes https://github.com/Liryna/pyfilesystem without implementing it
I was wondering where does the dll import append ? (python newbie)

```
from ctypes import *

try:
    DokanMain = windll.Dokan.DokanMain
    DokanVersion = windll.Dokan.DokanVersion
except AttributeError:
    raise ImportError("Dokan DLL not found")
```
Where does the dllimport happen ? How to set library name ?

**Comment by arekbulski** [comment]
You import windll from ctypes, which is a library loader. Member name is used as library name that is DLL name to be loaded. And for the library object, member name is used as function name. Simple example:

```
In [13]: ctypes.windll
Out[13]: <ctypes.LibraryLoader at 0x247b7dcdc88>
In [11]: ctypes.windll.msvcrt
Out[11]: <WinDLL 'msvcrt', handle 7ff829220000 at 0x247b8359f98>
In [12]: ctypes.windll.msvcrt.memset
Out[12]: <_FuncPtr object at 0x00000247B829FEE8>
```

**Comment by Liryna** [comment]
Thank you for your answer @arekbulski 
So `windll.dokan1` should load `dokan1.dll` ?

**Comment by arekbulski** [comment]
I think so, one way to know for sure.

**Comment by Liryna** [comment]
Yes I will make the test :smiley:
I don't have access to my dev environment so I am just reviewing the code for now.

**Comment by Liryna** [comment]
I am currently able to start the mount of a dokan device with my last changes https://github.com/Liryna/pyfilesystem
I see that `Mounted` operation is called, the python is correctly called but I get the error `run-time check failure the value of ESP`
I understood that this error means that the mapping python <-> C is not correct but I am unable to find what I did wrong. :( (first time I do this :smile: )

If someone could guide me, it would be nice !

https://github.com/dokan-dev/dokany/blob/master/dokan/dokan.h#L111
https://github.com/Liryna/pyfilesystem/blob/master/fs/expose/dokan/libdokan.py#L112

**Comment by arekbulski** [comment]
Throw it at Stack Overflow. I am not familiar enough with it.

**Comment by Liryna** [comment]
Probably in case USHORT was not define in earliest version of ctypes.
I removed it and forced c_ushort, same behaviour :cry: 
@lurch @zommuter are you familiar with ctypes ?

**Comment by arekbulski** [comment]
Not surprised. There is no point in supporting dinosaur-old versions of Python.

**Comment by Liryna** [comment]
My bad, I found that all call work with dokan1.dll in Release mode and not in Debug as I was doing.
Now I get proper call from dokan.

CreateFile fail here https://github.com/Liryna/pyfilesystem/blob/master/fs/expose/dokan/__init__.py#L453 when `\` directory is requested 
```
execpt Path contains invalid characters: \
```

I am not familiar with fs api and the code seems to not have been updated since 3 years. It there one of the pyfilesystem that could help me to review fs api calls ?

**Comment by arekbulski** [comment]
Here is the isdir() from MemoryFS and normpath() that it uses. From what I read in these methods, fs objects require/expect paths formatted in their way, so try some paths like "file1" or "/file1". 

Found it, I think. Notice that normpath does not recognize backslash at all. It is not recognized as one of ('..', '.', '') so it gets passed as a component, normpath('\') should give '/' right?.

https://github.com/Liryna/pyfilesystem/blob/master/fs/memoryfs.py#L321
https://github.com/Liryna/pyfilesystem/blob/master/fs/path.py#L20
https://github.com/Liryna/pyfilesystem/blob/master/fs/path.py#L66

**Comment by Liryna** [comment]
:+1: @arekbulski You are exactly right
normpath('\') is giving '\' :cry: 

I had to add in all functions:
path = path.replace('\\', '/')

I don't know if this is the best way but I am now able to run the test.

I am just not able to write the test file and will look into this.

**Comment by Liryna** [comment]
Pull request created: https://github.com/PyFilesystem/pyfilesystem/pull/258

I am able to Create/Delete/List/Rename/Read/Write files without issues
As exemple, Word 2016 successfully to open a file, read/write and save.

**Comment by arekbulski** [comment]
Well, it is you who is doing all the hard work. I am just good at code analysis. :smiley:

**Comment by arekbulski** [comment]
Just reviewed the pull request changes, posted as code review annotations. Btw, I reaally like that feature.

**Comment by Liryna** [comment]
A pleasure :+1: I always wanted dokan being compatible with python.

**Comment by arekbulski** [comment]
So what is the next thing forward?

**Comment by arekbulski** [comment]
I started reviewing expose/fuse and will fix some issues with It, but I would like to have an Active maintainer first.

**Comment by lurch** [comment]
> normpath('\') is giving '\'

That's correct - as far as pyfilesystem is concerned, a file named `\` is totally fine. 'input paths' to pyfilesystem should always be separated only by `/` characters ( http://docs.pyfilesystem.org/en/latest/concepts.html#paths ), and any conversion to/from backslashes (to handle the underlying OS / API calls) should be done inside the FS code. Have a look at OSFS to see if that helps?

**Comment by Liryna** [comment]
You are right `self.fs.isdir` does raise the error at some point with `\` as input path.
Does that mean OSFS should be fixed to handle such case ?
I will remove the replace if thats the case.

**Comment by lurch** [comment]
> Does that mean OSFS should be fixed to handle such case ?

No. As I already tried to explain, pyfilesystem has the concept of (abstract) "pyfilesystem paths", and (concrete) "system paths".
A pyfilesystem path might be something like `/some/dir/file.txt` and on Linux this might map to a (OSFS) system path of `/home/lurch/testing/some/dir/file.txt`, and on Windows it might map to a (OSFS) system path of `C:\Users\Andrew\Documents\testing\some\dir\file.txt`.
The user-code talking to pyfilesystem should only ever use forward-slashes to separate directories, and it's up to the library-code inside pyfilesystem to do any necessary manipulations to convert the pyfilesystem-paths to and from system-paths as necessary.

It's kinda difficult to explain clearly, but it means that user-code making use of pyfilesystem is always identical, no matter what platform it's running on, and what the backend filesystem is. (and this is also why pyfilesystem has no 'current directory' concept, because some of the backend filesystems don't support that)

**Comment by Liryna** [comment]
All right  :+1: , so If I understand correctly
my wait to convert path coming from Dokan CreateFile/Cleanup/Close operation like `\`, `\folder`, \folder\file.txt` to "pyfilesystem paths" with path.replace('\\', '/') is correct ?

**Comment by arekbulski** [comment]
I totally understand and agree with the concept behind abstract paths. However on Windows backslashes must be converted to a recognized separator or path splitting will lead to incorrect results.

**Comment by lurch** [comment]
I'm afraid I've never looked at any of the 'expose' classes in pyfilesystem, so maybe I'm getting confused between which paths are pyfilesystem-paths, and which paths are system-paths? :-S

But anyway, IMHO it'd be much cleaner to have just a couple of functions that convert Dokan-paths to and from pyfilesystem-paths, and use those in each of the other functions; rather than sprinkling `path.replace` calls everywhere ;)

Given that you've updated all the docstrings to say that the Dokan class now supports mounting to an arbitrary path, rather than (just) a drive-letter, have you tested that? Does there need to be any code adding and removing this extra path-info from the Dokan-paths? (apologies if that's a stupid question, but I've never looked at dokan code)

**Comment by arekbulski** [comment]
Agreed, should extract method to convert paths.

**Comment by Liryna** [comment]
I have added the extract method to convert paths.

**Comment by lurch** [comment]
Should this issue be closed, and progress tracked in #258 instead?

**Comment by arekbulski** [comment]
No, let is use this Issue as a floor for free discussion and PR for specific code changes. Could you change the title to something more like "Dokan main thread"?

**Comment by lurch** [comment]
@Liryna Have you tested mounting to a path instead of to a drive? (see https://github.com/PyFilesystem/pyfilesystem/pull/258#discussion-diff-72960386 and https://github.com/PyFilesystem/pyfilesystem/pull/241#issuecomment-170021043 ). Could you update the docstrings to clarify how it works?

**Comment by Liryna** [comment]
doc updated with an example.
Yes, this has been tested. It make no difference for the python. everything is transparent.

**Comment by lurch** [comment]
Thanks for the doc-update :+1:

**Comment by Liryna** [comment]
Et voilà ! :smiley: Added space

**Comment by arekbulski** [comment]
@Liryna Could you look at these issues and advise what to do about them? I am starting to see you as our main and only Dokan maintainer, if that is alright? I will try to work on those too.
#159 #175 #150 #149 #110

**Comment by Liryna** [comment]
@lurch could you make a full review of the pull request in one time ? This will make us win some time.
Keep in mind that this pull request add support of dokan 1.0.0.

Further changes should be made in another contribution.

**Comment by Liryna** [comment]
Can be close https://github.com/PyFilesystem/pyfilesystem/issues/159,  https://github.com/PyFilesystem/pyfilesystem/issues/150
Not dokan https://github.com/PyFilesystem/pyfilesystem/issues/175
Never seen this behaviour and probably python related https://github.com/PyFilesystem/pyfilesystem/issues/149, https://github.com/PyFilesystem/pyfilesystem/issues/110

**Comment by arekbulski** [comment]
Sorry, I am too sleepy and tired to read clearly. Pardon (trying french).
Ping @lurch Lets close some!

**Comment by lurch** [comment]
> @lurch could you make a full review of the pull request in one time ? This will make us win some time.

Sorry, I was just commenting on things as and when I saw them. I wasn't expecting you to react to my comments so quickly! ;-)

**Comment by lurch** [comment]
Closed a couple of issues (thanks!), but will leave the others open pending further testing.

@Liryna If I *do* get round to testing your updated Dokan support, and it all seems to work okay, do you mind if I wait until Dokan 1.0 is officially released before merging #258 , rather than relying on the Dokan RC version?

**Comment by arekbulski** [comment]
I think there is no point. Versioning schemes used in practice are mumbo jumbo anyway. RC is just another way of saying _stable but had major changes_. Same goes for 0.9 not being less stable than 1.0. Those are just fancy numbers. Tell me, am I wrong on this?

**Comment by lurch** [comment]
To me, RC signifies ReleaseCandidate, and implies "this is very close to the final code, so please help us test it, and we'll be releasing the real final version very soon". I'd rather wait until Dokan is non-RC, so that people (i.e. users of pyfilesystem wanting to experiment with Dokan) will be installing the final release of Dokan, rather than the soon-to-be-replaced RC version.
I have no concerns about stability, as I assume Dokan being on RC4 means that most of the major bugs have been found and fixed already.

**Comment by Liryna** [comment]
Current pyfilesystem implementation of dokan is only compatible with dokan
0.5.6.
I am even not aware if there is still somewhere a valid link to download
this version. Even if someone found it, it is by far unstable and probably
not correctly working on current windows version.

As you whish to keep this implementation until dokan 1.0.0 is released as
stable. Current RC and the next stable will have only fix reported by the
community but API will remain as it is.

Le 2 août 2016 1:15 AM, "Andrew Scheller" <notifications@github.com> a
écrit :

**Comment by lurch** [comment]
It's not that I "don't trust" the current RC version of Dokan, but surely it'd be better to advise end-users of pyfilesystem to just go and install https://github.com/dokan-dev/dokany/releases/latest (which is what https://github.com/dokan-dev/dokany/wiki/Installation recommends) rather than expecting them to seek out whatever the latest pre-release version is?

Yes, I'm aware the current Dokan support in pyfilesystem is horribly broken, which is why I'm looking forward to being able to merge this (hopefully) working version :-)  How far away is the final release of Dokan stable?

**Comment by arekbulski** [comment]
Liryna made the same point. Code we are working on will work with past-RC versions as well. There is no point in keeping it back.

**Comment by arekbulski** [comment]
@Liryna How about you mention what Lurch is talking about in the docstring? "Latest dokan binary needs to be installed, RC or otherwise."

**Comment by apocalypse2012** [comment]
As an end user, I would prefer access. Document the dependency details, certainly. But it is a reasonable assumption that an RC has a stable interface at least. It is easy for me to understand that the docker version required is not production ready, but I still need to develop my code in the meantime. Right now, I can't.

**Comment by arekbulski** [comment]
It is not an assumption that iterface will change, Liryna confirmed it.

**Comment by Liryna** [comment]
I think it is better to give the possibility to a end user to use
pyfilesystem with 1.0.0 RC or futur stable rather than the current
implementation that can absolutely not used since as explained before,
there is no way to find the dokan  version 0.5.3.

The futur stable will normally be released this month depending of the
community feedback that currently seems to be good!

Le 2 août 2016 11:14 AM, "Andrew Scheller" <notifications@github.com> a
écrit :

**Comment by Liryna** [comment]
Interface will not change at all. This pull request will remain as it is
even after stable release.

Le 2 août 2016 5:59 PM, "apocalypse2012" <notifications@github.com> a
écrit :

**Comment by lurch** [comment]
Fair enough, looks like I've been outvoted \*shrug\*
As I already said, I wasn't expecting the interface or the pyfilesystem-side of Dokan to change at all before or after RC / final versions, merely the Dokan download link. (I guess @apocalypse2012 shows I was wrong in assuming that people would prefer to install the stable version of Dokan, rather than a pre-release version)

I'm happy to go with the majority decision and go ahead and merge #258 anyway, once I've tested it as working. @Liryna have you tested it with different versions of Python, and on different versions of Windows?

**Comment by Liryna** [comment]
I only tested with python 2.7 for now and Windows 10.  Otherwise than
python limitations that I would not be aware, it should work on every
systems where dokan is compatible.

Le 2 août 2016 11:39 PM, "Andrew Scheller" <notifications@github.com> a
écrit :

**Comment by apocalypse2012** [comment]
What is my next step on this? Do I need to pull this from Github or is there a new package going up on PyPi?
Raymond 

    On Wednesday, August 3, 2016 12:20 AM, Liryna <notifications@github.com> wrote:
 

 I only tested with python 2.7 for now and Windows 10. Otherwise than
python limitations that I would not be aware, it should work on every
systems where dokan is compatible.

Le 2 août 2016 11:39 PM, "Andrew Scheller" <notifications@github.com> a
écrit :

**Comment by arekbulski** [comment]
Please edit your post. Those auto attached quotes are spamming my little, 5.5 inch screen. :)

**Comment by Liryna** [comment]
You can clone my pull request if you wanna test it before it is accepted.

Le 4 août 2016 7:01 PM, "apocalypse2012" <notifications@github.com> a
écrit :

**Comment by Liryna** [comment]
@apocalypse2012 have you been to give a try ?

**Comment by apocalypse2012** [comment]
Sadly Liryna, no. I have to admit, I have limited github experience and don't know how to do a pull from pyfilesystem or dokany to get at the version I need. Unfortunately, my plate has been too full to get it sorted. If you have any quick pointers to shortcut me through the process, they would be much appreciated...
Raymond 

    On Thursday, August 11, 2016 12:40 PM, Liryna <notifications@github.com> wrote:
 

 @apocalypse2012 have you been to give a try ?—

**Comment by Liryna** [comment]
git clone https://github.com/Liryna/pyfilesystem.git
cd pyfilesystem
python setup.py install

Pyfs doc http://docs.pyfilesystem.org/en/latest/getting_started.html#installing

You will probably need to remove package previously installed with pip.

**Comment by apocalypse2012** [comment]
Should the same work for Dokany to get the 1.0RC?

    On Thursday, August 11, 2016 2:00 PM, Liryna <notifications@github.com> wrote:
 

 git clone https://github.com/Liryna/pyfilesystem.git
cd pyfilesystem
python setup.py installPyfs doc http://docs.pyfilesystem.org/en/latest/getting_started.html#installingYou will probably need to remove package previously installed with pip.—

**Comment by Liryna** [comment]
direct link to download it: https://github.com/dokan-dev/dokany/releases/download/v1.0.0-RC4/DokanSetup_redist.exe

**Comment by Liryna** [comment]
@lurch have you been able to test the pull request ?

**Comment by lurch** [comment]
Not yet... as I mentioned in https://groups.google.com/forum/#!topic/pyfilesystem-discussion/0n7EGGIT0fo I'm actually away from home all this week.

**Comment by JokerQyou** [comment]
I gave it a try just now, with a Python 2.7.11 (32 bit) on Windows 10 64 bit, and the mounting subprocess went wrong, Windows shows a dialog saying `python.exe has stopped working`. And the traceback is as follows:

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "fs\expose\dokan\__init__.py", line 973, in mount
    check_ready(mp)
  File "fs\expose\dokan\__init__.py", line 943, in check_ready
    check_alive(mp)
  File "fs\expose\dokan\__init__.py", line 931, in check_alive
    raise OSError("dokan mount process exited prematurely")
OSError: dokan mount process exited prematurely
```

The code I tried is:

```python
import fs.memoryfs
import fs.expose.dokan
fs.memoryfs.__file__
s = fs.memoryfs.MemoryFS()
mp = fs.expose.dokan.mount(s, 'Q:\\')
```

My dokan installation is `Dokoan Library 1.0.0.4000 Bundle` as shown in Windows control panel.

I'm not sure whether it's a problem with my dokan installation or with your code. If I get a chance to reboot my computer I'll test again and let you know the result.

**Comment by Liryna** [comment]
Hi @JokerQyou ,

Did you copy the dokan1.dll from x86 folder (C:\Program Files\Dokan\x86) ?

**Comment by JokerQyou** [comment]
I tested again with `dokan1.dll` copied to the `pyfilesystem` project root folder, and it did not work as well.

```python
>>> import fs.memoryfs
>>> import fs.expose
>>> s = fs.memoryfs.MemoryFS()
>>> mp = fs.expose.
>>> from fs.expose import dokan
>>> mp = dokan.mount(s, 'Q:')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "fs\expose\dokan\__init__.py", line 923, in mount
    _check_path_string(path)
  File "fs\expose\dokan\__init__.py", line 896, in _check_path_string
    raise ValueError("invalid path: %r" % (path,))
ValueError: invalid path: 'Q:'
invalid path: 'Q:'
>>> mp = dokan.mount(s, 'Q:\\')
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "fs\expose\dokan\__init__.py", line 1061, in _do_mount
    mount(fs, path, **opts)
  File "fs\expose\dokan\__init__.py", line 964, in mount
    res = libdokan.DokanMain(ctypes.byref(opts), ctypes.byref(opstruct))
WindowsError: exception: access violation reading 0x00DF0000
KeyboardInterrupt
```

And I tested again with no mistake typing letters.

```python
>>> import fs.memoryfs
>>> from fs.expose import dokan
>>> s = fs.memoryfs.MemoryFS()
>>> mp = dokan.mount(s, 'Z:\\')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "fs\expose\dokan\__init__.py", line 973, in mount
    check_ready(mp)
  File "fs\expose\dokan\__init__.py", line 940, in check_ready
    check_alive(mp)
  File "fs\expose\dokan\__init__.py", line 931, in check_alive
    raise OSError("dokan mount process exited prematurely")
OSError: dokan mount process exited prematurely
dokan mount process exited prematurely
```

And this time the `dokan.mount()` takes very long to finish (with an error though), and the drive `Z:` did appear in explorer, but it says drive disconnected when I try to open it.

I was testing within an interactive shell, does that affect anything?

**Comment by Liryna** [comment]
```
>>> mp = dokan.mount(s, 'Q:\\')
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "fs\expose\dokan\__init__.py", line 1061, in _do_mount
    mount(fs, path, **opts)
  File "fs\expose\dokan\__init__.py", line 964, in mount
    res = libdokan.DokanMain(ctypes.byref(opts), ctypes.byref(opstruct))
WindowsError: exception: access violation reading 0x00DF0000
```
That's very strange

Can you try to run the mirror.exe in `C:\Program Files\Dokan\sample` ? (you need to enable developement files option during dokan installation)
`mirror.exe /r C:\Users /l m`

>I was testing within an interactive shell, does that affect anything?

This should not affect dokan.

If I understood correctly it mount the device but something is making it umount.
I have not been able to reproduce this so we should first look on your side if it come from the python or the library (by runing the mirror)

**Comment by Liryna** [comment]
@lurch have you been able to test it  ?

**Comment by lurch** [comment]
> @lurch have you been able to test it ?

Unfortunately not yet, been far too busy with other stuff :-(
It's still on my TODO list though...

**Comment by Liryna** [comment]
Dokan 1.0.0 has been released https://github.com/dokan-dev/dokany/releases/tag/v1.0.0
and can be used with my PR :wink:

**Comment by Liryna** [comment]
@JokerQyou I made a change on the pull request, do you think can you clone again and try to test it ?

**Comment by apocalypse2012** [comment]
Sorry for not having followed up on this. I've been impacted by a layoff that has taken me offline from my original motivation. Still interested in this library and will Follow up as soon as I can...

Raymond

**Comment by JokerQyou** [comment]
I'd like to give it a try but I've been too busy... And I cannot afford a system restart (I reinstalled dokan with development files but have not restarted my computer) now. I'll test it if I got time tonight (GMT+8).

**Comment by arekbulski** [comment]
I moved on to [Construct](https://github.com/construct/construct). Unsubscribing from this thread. Ping me if needed. :wave:

**Comment by lurch** [comment]
Moving this comment to here to try to separate 'general discussion' from 'code comments'...

> @lurch After 2 month, It would be nice if you could clone the PR and make your own opinion.

Sorry, I'm not sure what you're asking me for?

I **do** appreciate all the hard work you and others have been putting into getting this working :+1:  I've never written Win32 API code myself, but I understand how 'murky' it can be.
Unfortunately, if even Windows Explorer or Notepad are currently unable to save files into a dokan-exposed filesystem (which AFAICT is the whole reason for the 'expose' interface in the first place - right @willmcgugan ?) then IMHO #258 has limited usefulness for typical users. People would only complain when it doesn't work the way they expect it to.

OTOH if you wanted to modify the PR so that it can _only_ be used to expose ReadOnly filesystems (see e.g. http://docs.pyfilesystem.org/en/latest/base.html#fs.base.FS.getmeta and http://docs.pyfilesystem.org/en/latest/wrapfs/readonlyfs.html ), and documented it as such, then that _could_ probably be merged. (Which will then hopefully be 'upgraded' later on to allow full dokan read/write access when this security issue gets fixed)
It would be limited functionality (no modification to the exposed FS), but at least it would be consistent behaviour for the end-user.

I hope that doesn't sound too harsh, I hope you can see where I'm coming from. I'm afraid I don't get as much time to look at pyfilesystem as I'd like, and from their lack of input I assume @willmcgugan and @rfk are even busier than I am :-/

**Comment by willmcgugan** [comment]
Thanks @lurch. Sorry I haven't been keeping up with the work here.

Not being a massive Windows user I'll need a little guidance here. If we're talking about fixes then I'm all for it. But I'm guessing its not as clear cut?

**Comment by JokerQyou** [comment]
I've managed to test directly with the whole repo cloned from @Liryna 's remote, and here's the result with memory fs.

```python
λ ptpython                                                                       
>>> import fs.memoryfs                                                           
>>> import fs.expose                                                             
>>> fs                                                                           
<module 'fs' from 'fs\__init__.py'>                                              
>>> s = fs.memoryfs.MemoryFS()                                                   
>>> from fs.expose import dokan                                                  
>>> mp = dokan.mount(s, 'Q:')                                                    
Traceback (most recent call last):                                               
  File "<stdin>", line 1, in <module>                                            
  File "fs\expose\dokan\__init__.py", line 957, in mount                         
    _check_path_string(path)                                                     
  File "fs\expose\dokan\__init__.py", line 929, in _check_path_string            
    raise ValueError("invalid path: %r" % (path,))                               
ValueError: invalid path: 'Q:'                                                   
invalid path: 'Q:'                                                               
>>> mp = dokan.mount(s, 'Q:\\')                                                  
>>> mp                                                                           
<fs.expose.dokan.MountProcess object at 0x050FEBB0>                              
>>> s                                                                            
MemoryFS()                                                                       
>>> s.listdir()                                                                  
[]                                                                               
>>> s.listdirinfo()                                                              
[]                                                                               
```

At this time `Q:` is mounted successfully and can is visible through explorer, and I've created a folder in it (using explorer's context menu), but the created folder does not show up with `s.listdir()`.

And I've also tried to created a new folder with `s.makedir('a')`, this folder showed up with `s.listdir()`, but not in explorer. The mounted drive in explorer does not sync up with the memory fs in python.

And I used the following code to unmount the drive:

```python
>>> dokan.unmount('Q:\\')
>>> Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "fs\expose\dokan\__init__.py", line 1096, in _do_mount
    mount(fs, path, **opts)
  File "fs\expose\dokan\__init__.py", line 998, in mount
    res = libdokan.DokanMain(ctypes.byref(opts), ctypes.byref(opstruct))
WindowsError: exception: access violation reading 0x00000010
```

This error was not raised right after my command returned. Actually the unmount method returned very fast, but it was after several more seconds did the error come up. I suspect it was the background process which crashed.

For pyfilesystem it was a fresh clone from https://github.com/Liryna/pyfilesystem.git , as for Dokan it was `1.0.0.4000` (verfion of dokan1.dll). I'll try the newer version of Dokan some time later.

**Comment by Rondom** [comment]
Current master in this repo should have all necessary changes. If memory serves me right, dokan.unmount is a callback that is called when the volume is unmounted.

**Comment by BiatuAutMiahn** [comment]
I can't get this to work, Dokany v1.3.0.1000

Output:
```
>>> from fs.memoryfs import MemoryFS
>>> from fs.expose import dokan
>>> fs = MemoryFS()
>>> # Mount device mount point
... mp = dokan.mount(fs, "Q:\\")
Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "C:\ProgramData\Miniconda3\lib\site-packages\fs\expose\dokan\__init__.py", line 1107, in _do_mount
    mount(fs, path, **opts)
  File "C:\ProgramData\Miniconda3\lib\site-packages\fs\expose\dokan\__init__.py", line 1011, in mount
    raise OSError("Dokan failed with error: %d" % (res,))
OSError: Dokan failed with error: -7
Traceback (most recent call last):
  File "C:\ProgramData\Miniconda3\lib\site-packages\fs\expose\dokan\__init__.py", line 983, in check_ready
    os.stat(path)
FileNotFoundError: [WinError 3] The system cannot find the path specified: 'Q:\\'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "C:\ProgramData\Miniconda3\lib\site-packages\fs\expose\dokan\__init__.py", line 1018, in mount
    check_ready(mp)
  File "C:\ProgramData\Miniconda3\lib\site-packages\fs\expose\dokan\__init__.py", line 985, in check_ready
    check_alive(mp)
  File "C:\ProgramData\Miniconda3\lib\site-packages\fs\expose\dokan\__init__.py", line 976, in check_alive
    raise OSError("dokan mount process exited prematurely")
OSError: dokan mount process exited prematurely
```

**Comment by Liryna** [comment]
Will take a look when I can. Otherwise are you sure to use same 32/64
platform for dokan dll and python?

Le sam. 31 août 2019 à 23:09, BiatuAutMiahn <notifications@github.com> a
écrit :

**Comment by BiatuAutMiahn** [comment]
Tried both, CWD was in the dokany dll folder

**Comment by BiatuAutMiahn** [comment]
I ran dokanctl  to see options, enabled debug then found DebugView, tried to run `mp = dokan.mount(fs, "Q:\\")` again and I found this...
![image](https://user-images.githubusercontent.com/6149596/64455944-f6874500-d0bc-11e9-8c6a-3526448b86be.png)

**Comment by Liryna** [comment]
Oh this is helpful!

There is a version number in dokan py file normal. You would need to update
the value but this also mean probably something in the python wrapper need
to be updated also.
Will try to take a look in the next weeks if you want.

Best regards
Adrien JUND

Le ven. 6 sept. 2019 à 21:42, BiatuAutMiahn <notifications@github.com> a
écrit :

**Comment by BiatuAutMiahn** [comment]
Changed:
`DOKAN_MINIMUM_COMPATIBLE_VERSION = 100  # this is release 1.0.0` 
to
`DOKAN_MINIMUM_COMPATIBLE_VERSION = 110  # this is release 1.0.0`

and Q:\ has mounted!

**Comment by Rondom** [comment]
> There is a version number in dokan py file normal. You would need to update
the value **but this also mean probably something in the python wrapper need
to be updated also.**

You need to a diff between dokan.h of 1.0.0 and latest Dokan and adapt the Python code accordingly. Otherwise you will get random crashes (which is why there are those version checks).

**Comment by vpuhoff** [comment]
https://github.com/nznaza/pyfilesystem_dokan

**Comment by Liryna** [comment]
If anyone would be interested, dokan-dev is still looking for an official dokan python wrapper!
There is already some code base available here and there, it might not be much work to put everything together and make it available to the community 👍 

Ping @nznaza

**Comment by nznaza** [comment]
I have finally decided to look back into my repo. (When I first uploaded it I got burned out, and studies and whatnot got in the way)

Creating, Deleting and Renaming Folders work.
Copying Files works.
Creating, Deleting and Modifying files works too,
Moving files to another drive works too.
Moving/renaming files it's the only major bug that I noticed. I'm looking into it.
I have tested it with MemoryFS, but since it relies in PyFilesystem any should work.

If people test and report bugs I will happily look into them.

**Comment by hugo-hur** [comment]
Hi!

Any updates on this? Looking to use this to mount ramfs on Windows to display decrypted files using Python3.

**Comment by hugo-hur** [comment]
Could be willing to help with this, but would it be better suited here or the Dokany repo? Imho would be simpler to write an intermediate wrapper using C/C++ and then create python bindings to that intermediate wrapper optimized to use with Python?
@Liryna

**Comment by hugo-hur** [comment]
And then most of the work would be associated with updating the C/C++ intermediate binding interface?

**Comment by Liryna** [comment]
@hugo-hur 
@kdschlosser Started a new python wrapper for Dokan https://github.com/kdschlosser/py_dokany
From his last comment https://github.com/dokan-dev/dokany/issues/964#issuecomment-774325735 it looks like it is working and probably need to be finished / polished. Unfortunately, there was no update after that so we would need someone else to take over 💪

**Comment by kdschlosser** [comment]
No matter what I did it would not enumerate a directory properly. .. the sheer number duplicate requests was also an issue and made it not visible to use for what I had wanted to use it for.
